### PR TITLE
Use DMA APIs to refactor virtio drivers

### PR DESCRIPTION
**Pull request by StevenJiang1110**

### CODE REVIEW

**Comment by tatetian** [comment]
@liqinggd Could you review this PR?

File: kernel/comps/network/src/dma_pool.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            allocated_blocks: SpinLock::new(BitArray::ZERO),
+        });
+
+        let block = dma_page.alloc_block(self.block_size).unwrap();
+        self.pages.write_irq_disabled().push_back(dma_page);
+        Ok(block)
+    }
+
+    /// Free pages that are not used in the pool
+    pub fn free_pages(&self) {
```
Comment (liqinggd) [inline kernel/comps/network/src/dma_pool.rs:159]: The API `alloc_block` and `free_pages` makes the users confused.

File: kernel/comps/network/src/dma_pool.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    fn drop(&mut self) {
+        let page = self.page.upgrade().unwrap();
+        debug_assert!((page.daddr()..page.daddr() + PAGE_SIZE).contains(&self.daddr()));
+        let block_idx = (self.daddr() - page.daddr()) / self.size;
+        page.allocated_blocks.lock().set(block_idx, false);
+    }
+}
+
+impl DmaPool {
+    pub fn new(block_size: usize, direction: DmaDirection, is_cache_coherent: bool) -> Self {
```
Comment (liqinggd) [inline kernel/comps/network/src/dma_pool.rs:118]: I think the `block_size` can be a `const genetic`. This is because the size of `DmaBlock` is actually determined at the time when `DmaPool` is created, and this size will not change afterwards.

File: kernel/comps/network/src/dma_pool.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+#[derive(Debug)]
+pub struct DmaPool {
+    direction: DmaDirection,
+    block_size: usize,
+    is_cache_coherent: bool,
+    pages: RwLock<VecDeque<Arc<DmaPage>>>,
+}
+
+#[derive(Debug)]
+struct DmaPage {
```
Comment (liqinggd) [inline kernel/comps/network/src/dma_pool.rs:33]: `DmaPage` must ensure that its size is equal to `PAGE_SIZE`.

File: kernel/comps/network/src/buffer.rs [hunk 4]
```diff
@@ -33,59 +93,49 @@ impl RxBuffer {
     }
 
     pub fn set_packet_len(&mut self, packet_len: usize) {
-        self.packet_len = packet_len;
-    }
-
-    pub fn buf(&self) -> &[u8] {
-        &self.buf
+        self.packet_len = packet_len
```
[the remark this answers is not in the document]
Comment (StevenJiang1110) [reply kernel/comps/network/src/buffer.rs]: A new check is added to `new` and `set_packet_len` now.

File: kernel/comps/virtio/src/device/console/device.rs [hunk 5]
```diff
@@ -43,17 +59,26 @@ impl AnyConsoleDevice for ConsoleDevice {
...
         let (_, len) = receive_queue.pop_used().unwrap();
+        self.receive_buffer.sync(0..len as usize).unwrap();
+
+        let recv_len = {
+            // FIXME: If the length of the recv_buffer exceeds that of the buf,
+            // there will be some content that is not received.
+            let mut writer = VmWriter::from(buf);
+            let mut reader = self.receive_buffer.reader().unwrap().limit(len as usize);
+            writer.write(&mut reader)
+        };
```
[the remark this answers is not in the document]
Comment (StevenJiang1110) [reply kernel/comps/virtio/src/device/console/device.rs]: After reviewing the current codebase, the `recv` function is never used. Tty device will register callbacks on virtio-console instead of calling `recv`. I have removed the `recv` method.

File: kernel/comps/virtio/src/device/input/device.rs [hunk 6]
```diff
@@ -181,44 +185,133 @@ impl InputDevice {
...
             let callbacks = self.callbacks.lock();
             for callback in callbacks.iter() {
-                callback.call((event,));
+                callback(event);
             }
+
+            true
+        };
+
+        // one interrupt may contains serval input, so it should loop
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/input/device.rs:221]: Typo: "serval input"

File: kernel/comps/virtio/src/device/input/device.rs [hunk 7]
```diff
@@ -181,44 +185,133 @@ impl InputDevice {
...
                 1 => KeyStatus::Pressed,
                 0 => KeyStatus::Released,
-                _ => return Some(()),
+                _ => return false,
             };
+
             let event = InputEvent::KeyBoard(Key::try_from(event.code).unwrap(), status);
             info!("Input Event:{:?}", event);
 
             let callbacks = self.callbacks.lock();
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/input/device.rs:213]: This callback list is mostly read. Protect it with a read-write lock. Apply this optimization to all driver that contains a callback list in a separate commit.

File: kernel/comps/virtio/src/device/input/device.rs [hunk 8]
```diff
@@ -181,44 +185,133 @@ impl InputDevice {
...
             for callback in callbacks.iter() {
-                callback.call((event,));
+                callback(event);
             }
+
+            true
+        };
+
+        // one interrupt may contains serval input, so it should loop
+        while self.pop_pending_event(&handle_event) {}
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/input/device.rs:222]: For optimal efficiency, it is best to avoid locking `callbacks` repeatedly in this function and `event_queue` in the `pop_pending_event` function. Consider refactoring both functions to the following form.

```rust
impl InputDevice {
    // Change note: Process events in batch, instead of one by one.
    fn pop_pending_events(&self, handle_event: &impl Fn(&EventBuf) -> bool) {
        let mut event_queue = self.event_queue.lock_irq_disabled();
        while let Ok((token, _)) = event_queue.pop_used() {
            // handle the event
        }
    }

    fn handle_irq(&self) {
        //  Change note: Lock the callback per IRQ, not per event.
        let callbacks = self.callbacks.read();

        let handle_event = |event: &EventBuf| -> bool {
            // ..

            for callback in callbacks.iter() {
                callback(event);
            }
        };

        // Change note: No loop here
        self.pop_pending_events(&handle_event);
    }
}
```

File: kernel/comps/virtio/src/device/input/device.rs [hunk 9]
```diff
@@ -181,44 +185,133 @@ impl InputDevice {
...
+        let vm_segment = {
+            let mut options = VmAllocOptions::new(1);
+            options.is_contiguous(true);
+            options.alloc_contiguous().unwrap()
+        };
+
+        let default_event = VirtioInputEvent::default();
+        for idx in 0..num_events {
+            let offset = idx * EVENT_SIZE;
+            vm_segment.write_val(offset, &default_event).unwrap();
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/input/device.rs:253]: Writing a value repeatedly seems to be a common need. Consider adding a new method called `writes_vals`.

```rust
impl VmIo {
    /// Write a sequence of values given by an iterator (`iter`) from the specified offset (`offset`).
    /// 
    /// The write process stops until the VM object does not have enough remaining space
    /// or the iterator returns `None`. If any value is written, the function returns `Ok(nr_written)`,
    /// where `nr_written` is the number of the written values.
    ///
    /// The offset of every value written by this method is aligned to the `align`-byte boundary.
    /// Naturally, when `align` equals to `0` or `1`, then the argument takes no effect:
    /// the values will be written in the most compact way.
    ///
    /// # Example
    ///
    /// Initializing an VM object with the same value can be done easily with `write_values`.
    ///
    /// ```
    /// use core::iter::self;
    ///
    /// let _nr_values = vm_obj.write_values(0, iter::repeat(0_u32), 0).unwrap();
    /// ```
    fn write_values<T: Pod, I: Iterator<Item = T>>(&self, offset: usize, iter: I, align: usize) -> Result<usize> {
        // ...
    }
}
```

This extension to `VmIo` should be added in a separate commit.

(I am not quite sure that the trait bound for `I` is the most appropriate one)

File: kernel/comps/virtio/src/device/input/device.rs [hunk 10]
```diff
@@ -181,44 +185,133 @@ impl InputDevice {
...
 
+    const fn num_events(&self) -> usize {
+        self.num_events
+    }
+}
+
+const EVENT_SIZE: usize = core::mem::size_of::<VirtioInputEvent>();
+
+/// A buffer stores exact one `VirtioInputEvent`
+struct EventBuf<'a> {
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/input/device.rs:279]: This `EventBuf` is every much like `SafePtr`. I am thinking if `SafePtr` can be extended to satisfy this need or a a new `DmaPtr` should be added.

Comment (StevenJiang1110) [reply kernel/comps/virtio/src/device/input/device.rs]: I have extended `SafePtr` to support `DmaStream`. Then EventBuf can be replaced with `SafePtr<VirtioInputEvent, DmaStream>`

```rust
impl<T, M: HasDaddr, R> HasDaddr for SafePtr<T, M, R> {
    fn daddr(&self) -> Daddr {
        self.offset + self.vm_obj.daddr()
    }
}

impl<T, R> SafePtr<T, DmaStream, R> {
    /// Synchronize the object in the streaming DMA mapping
    pub fn sync(&self) -> Result<()> {
        self.vm_obj
            .sync(self.offset..self.offset + core::mem::size_of::<T>())
    }
}
```

File: kernel/comps/virtio/src/device/input/device.rs (hunk 9 above)
Comment (StevenJiang1110) [reply kernel/comps/virtio/src/device/input/device.rs]: I have rewritten the signature as
```rust
fn write_vals<'a, T: Pod + 'a, I: Iterator<Item = &'a T>>(
        &self,
        offset: usize,
        iter: I,
        align: usize,
    ) -> Result<usize> 
```

Ideally, the `Item` of `Iterator` can be either `T` or `&T`. But I can't find a type to support both. So I use `&T` here.

**Comment by tatetian** [comment]
@StevenJiang1110 Could you also remove the `add_buf` method of virtio queue (and its helper functions, if any)? With the refactoring in this PR, there should be no users of this API.

File: kernel/comps/virtio/src/device/input/device.rs [hunk 11]
```diff
@@ -244,23 +242,19 @@ impl EventTable {
...
+    fn get(&self, idx: usize) -> EventBuf {
         assert!(idx < self.num_events);
 
         let offset = idx * EVENT_SIZE;
-        EventBuf {
-            event_table: self,
-            offset,
-            size: EVENT_SIZE,
-        }
+        SafePtr::new(self.stream.clone(), offset)
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/input/device.rs:257]: I think it should be ok to construct a safe pointer by borrowing the DMA stream, which is cheaper than cloning.

Comment (StevenJiang1110) [reply kernel/comps/virtio/src/device/input/device.rs]: I have Inherited `HasDaddr` for `&T` if `T` implements `HasDaddr`. So currently `EventBuf` becomes `SafePtr<VirtioInputEvent, &'a DmaStream>`, the `clone` is avoided.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
